crates/wasm-mutate/src/mutators/codemotion/ir/mod.rs

```diff
@@ -0,0 +1,377 @@
+use crate::{
+    module::map_block_type,
+    mutators::{codemotion::ir::parse_context::ParseContext, OperatorAndByteOffset},
+};
+use wasm_encoder::{Function, Instruction};
+use wasmparser::{Operator, Range, TypeOrFuncType};
+
+use self::parse_context::{Ast, Node, State};
+pub struct AstBuilder;
+pub mod parse_context;
+
+/// Encodes the Wasm Ast
+pub trait AstWriter {
+    /// Encodes a loop node.
+    ///
+    /// Redefine this if your implementation mutates loops. For example, if your
+    /// implementation unrolls the first iteration of a particular loop, override
+    /// this method. For the other loops in the Wasm that your implementation is not
+    /// mutating, you can call `write_loop_default`.
+    ///
+    fn write_loop<'a>(
+        &self,
+        ast: &Ast,
+        nodeidx: usize,
+        body: &[usize],
+        newfunc: &mut Function,
+        operators: &Vec<OperatorAndByteOffset>,
+        input_wasm: &'a [u8],
+        ty: &TypeOrFuncType,
+    ) -> crate::Result<()> {
+        self.write_loop_default(ast, nodeidx, body, newfunc, operators, input_wasm, ty)
+    }
+
+    /// Default encoding for a loop node
+    /// This function is called by the defaut implementation
+    /// of the `write_loop` method
+    ///
+    fn write_loop_default<'a>(
+        &self,
+        ast: &Ast,
+        _nodeidx: usize,
+        body: &[usize],
+        newfunc: &mut Function,
+        operators: &Vec<OperatorAndByteOffset>,
+        input_wasm: &'a [u8],
+        ty: &TypeOrFuncType,
+    ) -> crate::Result<()> {
+        newfunc.instruction(&Instruction::Loop(map_block_type(*ty)?));
+        for ch in body {
+            self.write(ast, *ch, newfunc, operators, input_wasm)?;
+        }
+        newfunc.instruction(&Instruction::End);
+        Ok(())
+    }
+
+    /// Default encoding for a loop node
+    /// This function is called by the defaut implementation
```

Comment:
It looks like you missed reformatting a couple doc comments on this trait's methods.